Add new feature `Spatial Bookmarks` from QGIS project (User and Project)

volterra79 opened the pull request:
Closes: #321 

---

Add a Sidebar item as follow image:

![Screenshot from 2023-02-08 09-49-11](https://user-images.githubusercontent.com/1051694/217480408-9f40e133-409f-4857-b8d2-65027b9b255c.png)

And show a list of Project and User Bookmarks as below:

![Screenshot from 2023-02-08 09-50-19](https://user-images.githubusercontent.com/1051694/217480684-49ec8ef3-b8d5-4c0e-b068-b38f849fe5bb.png)

For the User Bookmarks ad a (+) button to add a spatial bookmarks:

![Screenshot from 2023-02-08 09-52-27](https://user-images.githubusercontent.com/1051694/217481205-d8c18c74-5173-4605-a852-5d0c35e21a4e.png)


**For more info:**

- https://docs.qgis.org/3.22/en/docs/user_manual/introduction/general_tools.html#spatial-bookmarks

volterra79 committed:
Init implementation of Spatial BookMarks on sidebar

volterra79 committed:
sidebar change to improve and have a bind with action sidebar item with owner component

volterra79 committed:
Add emit handler when click on plus button to open sidebar items

volterra79 committed:
Merge branch 'dev' into v3.6_feature-spatial-bookmarks-sidebar

volterra79 committed:
Add getSpatialBookmarks to project to get eventually spatial bookmarks store on QGIS Project

volterra79 committed:
Merge branch 'dev' into v3.6_feature-spatial-bookmarks-sidebar

volterra79 committed:
Add SpatialBookmarkItem component

volterra79 committed:
Add g3w-long-text to handle input error message in small space

volterra79 committed:
Add LOCALSTORAGE_IDS sto have LocalStorage app items unique

volterra79 committed:
Add SpatialBookmarkGroup component

volterra79 committed:
Remove fake bookmarks from server

volterra79 committed:
Remove add bookmark button on sidebar item

volterra79 committed:
Remove add emit event

volterra79 committed:
User and Project Bookmark. Add Group and item bookmark child components

volterra79 committed:
Add i18n translations

volterra79 committed:
LOCALITEMSIDS on costant and initial value

Raruto commented on src/components/SpatialBookMarks.vue:
```diff
@@ -0,0 +1,177 @@
+<template>
```
```suggestion
<!--
  @file
  @since v3.8
-->

<template>
```

Raruto commented on src/components/SpatialBookMarkItem.vue:
```diff
@@ -0,0 +1,59 @@
+<template>
```
```suggestion
<!--
  @file
  @since v3.8
-->

<template>
```

Raruto commented on src/components/SpatialBookMarkGroup.vue:
```diff
@@ -0,0 +1,39 @@
+<template>
```
```suggestion
<!--
  @file
  @since v3.8
-->

<template>
```

Raruto commented on src/app/gui/spatialbookmarks/vue/spatialbookmarks.js:
```diff
@@ -0,0 +1,28 @@
+import * as vueComponentOptions from 'components/SpatialBookMarks.vue';
```
```suggestion
/**
 * @file
 * @since v3.8
 */
 
import * as vueComponentOptions from 'components/SpatialBookMarks.vue';
```

Raruto commented on src/app/core/project/project.js:
```diff
@@ -456,5 +457,13 @@ proto.getUrl = function(type){
   return this.urls[type];
 };
 
+/**
+ *
+ * @returns {<Array> spatail bookmarks saved on QGIS project}
+ */
```
```suggestion
/**
 * @returns {Array} spatial bookmarks saved on current QGIS project
 * @since v3.8
 */
```

Raruto commented on src/app/constant.js:
```diff
@@ -356,6 +356,24 @@ export const API_BASE_URLS = {
   CRS: '/crs/'  //Example /crs/<epsg_code>
 };
 
+/**
+ * LOCAL ITEM IDS
+ * Used to store id used by application
+ * Each item has:
+ * {
+ *  id: unique identifier,
+ *  value: initial value
+ *  }
+ * @since v3.8
+ */
+
```
```suggestion
/**
 * @typedef LocalItem
 * @property {string} id unique identifier
 * @property {any} value initial value
 */

/**
 * LOCAL ITEM IDS - used to store id used by application
 * 
 * @type {Object<string, LocalItem>}
 * 
 * @since v3.8
 */
```

Raruto commented on src/app/constant.js:
```diff
@@ -356,6 +356,24 @@ export const API_BASE_URLS = {
+ * @since v3.8
+ */
+
+export const LOCALITEMSIDS = {
```
Perhaps `LOCAL_ITEM_IDS` would be easier to read

volterra79 committed:
Update src/components/SpatialBookMarks.vue

Co-authored-by: Raruto <Raruto@users.noreply.github.com>

volterra79 committed:
Update src/components/SpatialBookMarkItem.vue

Co-authored-by: Raruto <Raruto@users.noreply.github.com>

volterra79 committed:
Update src/components/SpatialBookMarkGroup.vue

Co-authored-by: Raruto <Raruto@users.noreply.github.com>

volterra79 committed:
Update src/app/gui/spatialbookmarks/vue/spatialbookmarks.js

Co-authored-by: Raruto <Raruto@users.noreply.github.com>

volterra79 committed:
Update src/app/core/project/project.js

Co-authored-by: Raruto <Raruto@users.noreply.github.com>

volterra79 committed:
Update src/app/constant.js

Co-authored-by: Raruto <Raruto@users.noreply.github.com>

volterra79 committed:
Change LOCALITEMSIDS to LOCAL_ITEM_IDS as suggest https://github.com/g3w-suite/g3w-client/pull/322#discussion_r1101138485

Raruto committed:
Update missing exports within `index.js`

volterra79 committed:
Merge branch 'dev' into v3.6_feature-spatial-bookmarks-sidebar

volterra79 committed:
add force option parameter attribute to force to use passed extent for get map resolution

Raruto committed:
update jsdoc + refactor `getResolutionForZoomToExtent`

volterra79 commented on src/app/gui/map/mapservice.js:
```diff
@@ -2178,38 +2176,32 @@ proto.compareExtentWithProjectMaxExtent = function(extent){
-  return resolution
+
+  // if outside project extent, return max resolution
+  if (!ol.extent.containsExtent(this.project.state.extent, extent)) {
```
@Raruto we can write it as

```js
if (false === ol.extent.containsExtent(this.project.state.extent, extent))
```

volterra79 commented on src/app/gui/map/mapservice.js:
```diff
@@ -2178,38 +2176,32 @@ proto.compareExtentWithProjectMaxExtent = function(extent){
+  // retrive resolution from given `extent`
+  if (true === options.force) {
+    return extentResolution; 
+  }
```
@Raruto 

```js
  const extentResolution = map.getView().getResolutionForExtent(extent, map.getSize());            // resolution of request extent

  // retrive resolution from given `extent`
  if (true === options.force) {
    return extentResolution; 
  }

 const maxResolution    = getResolutionFromScale(MAP_SETTINGS.ZOOM.maxScale, this.getMapUnits()); // max resolution of the map
```

Raruto commented:
The error input message does not meet WCAG standards for accesible text (ref: [perceivable color contrast](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Perceivable/Color_contrast))

![image](https://user-images.githubusercontent.com/9614886/220573266-89d0f52f-c98b-4d83-8fbf-60b295e0f718.png)

---

The input field for adding a new user bookmark (bookmark name) is not cleared before creating a new one (as you can see `foo` is displayed before I add the `bar` bookmark):

![image](https://user-images.githubusercontent.com/9614886/220575772-5460643b-5f49-4a89-90a5-ea3ba1e97625.png)

![image](https://user-images.githubusercontent.com/9614886/220576313-2f01708e-c483-4a2f-9de6-befe0b26bd61.png)

volterra79 committed:
Add/change translation

volterra79 committed:
Change style of inner spatial book mark item component to have more readability between group and item

volterra79 committed:
Change color of input message to have more contrast.
Add showAddForm method to show form and set input text value to null asu suggest https://github.com/g3w-suite/g3w-client/pull/322#issuecomment-1439682499

volterra79 committed:
Merge branch 'dev' into v3.6_feature-spatial-bookmarks-sidebar

Raruto commented on src/components/SpatialBookMarks.vue:
```diff
@@ -0,0 +1,191 @@
+          class="container add-bookmark-input"
+          style="padding: 5px; width: 100%"
+        >
+          <input-text
+            :state="addbookmarkinput"/>
```
here I thought of adding something like `@keyup.enter="addBookMark"`, to be able to add a bookmark even just by pressing the enter key.

Do you think it's somehow feasible or it requires too many modifications elsewhere?

volterra79 replied:
@Raruto for the moment leave it as it is. Need to click on button to add bookmark

Raruto committed:
add translations

Raruto committed:
small refactor of `getResolutionForZoomToExtent`

Raruto commented:
A new feature has been delivered 🚀, great job!
